core/src/main/java/org/infinispan/partitionhandling/impl/PartitionHandlingInterceptor.java

```diff
@@ -129,20 +132,20 @@ private boolean performPartitionCheck(InvocationContext ctx, LocalFlagAffectedCo
    }
 
    private CompletableFuture<Void> handleDataReadCommand(InvocationContext ctx, DataCommand command) {
-      Object key = command.getKey();
       return ctx.onReturn((rCtx, rCommand, rv, throwable) -> {
+         DataCommand dataCommand = (DataCommand) rCommand;
          if (throwable != null) {
-            if (throwable instanceof RpcException && performPartitionCheck(rCtx, ((DataCommand) rCommand))) {
+            if (throwable instanceof RpcException && performPartitionCheck(rCtx, dataCommand)) {
                // We must have received an AvailabilityException from one of the owners.
                // There is no way to verify the cause here, but there isn't any other way to get an invalid
                // get response.
-               throw log.degradedModeKeyUnavailable(key);
+               throw log.degradedModeKeyUnavailable(dataCommand.getKey());
```

Comment:
True, asymmetric partitioning and even the behaviour of cache operations during a merge aren't tested in depth. That's exactly why I'm hesitating so much about the `ClusteredGetValidityFilter` changes...

AFAICT with this PR a node can become `AVAILABLE`, read a key from one node, but then fail to read a remote key because the primary owner is still `DEGRADED`. Without this PR, the 2nd read would only fail if all the owners of the key are `DEGRADED`, but it would still fail in some situations, because `ClusteredGetCommand` doesn't have a topology id and so it can't block until the remote nodes become available. In the end, the difference is quite small, but this PR is supposed to "fix RPC handling", so I wouldn't include a change that makes things (ever so slightly) worse.